File: daily_problems/2026/07/0729/personal_submission/cf106628l_yawn_sean.cpp
#include <bits/stdc++.h>
// #include "atcoder/all"
// #pragma GCC optimize("Ofast")
// #pragma GCC optimize("unroll-loops")
// #pragma GCC target("sse,sse2,sse3,ssse3,sse4,popcnt,abm,mmx,avx2,tune=native")

using namespace std;

auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

long long quickPow(long long base, long long power, long long mod) {
	if (power == 0) return 1 % mod;
	long long cur = quickPow(base, power / 2, mod);
	return power & 1 ? base * cur % mod * cur % mod : cur * cur % mod; 
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int k, mod = 1e9 + 7;
	cin >> k;

	vector<pair<int, int>> pts(k + 2);
	for (auto &[x, y]: pts) cin >> x >> y;

	vector<int> saved_vals(k + 2);

	for (int i = 0; i < k + 2; i ++) {
		int val = 1;
		for (int j = 0; j < k + 2; j ++) {
			if (i == j) continue;
			val = 1ll * val * (mod + pts[i].first - pts[j].first) % mod;
		}
		saved_vals[i] = pts[i].second * quickPow(val, mod - 2, mod) % mod;
	}

	for (int i = 0; i < k + 2; i ++) {
		int val = 0;
		for (int j = 0; j < k + 2; j ++) {
			if (i == j) continue;
			val += 1ll * saved_vals[j] * (mod + pts[i].first - pts[j].first) % mod;
			val %= mod;
		}
		if (val == 0) {
			int ans = 0;
			for (int j = 0; j < k + 2; j ++) {
				if (j != i) {
					int A = pts[j].second, B = 1;
					for (int v = 0; v < k + 2; v ++) {
						if (v != i && v != j) {
							A = 1ll * A * (mod - pts[v].first) % mod;
							B = 1ll * B * (mod + pts[j].first - pts[v].first) % mod;
						}
					}
					ans += A * quickPow(B, mod - 2, mod) % mod;
					ans %= mod;
				}
			}
			cout << ans << '\n';
			break;
		}
	}

	return 0;
}

File: daily_problems/2026/07/0729/personal_submission/cf106628l_yawn_sean.py
# Submission link: https://codeforces.com/gym/106628/submission/384575957
def main():
    mod = 10 ** 9 + 7
    k = II()
    
    xs = []
    ys = []
    
    for _ in range(k + 2):
        x, y = MII()
        xs.append(x)
        ys.append(y)
    
    saved_vals = [0] * (k + 2)
    
    for i in range(k + 2):
        val = 1
        
        for j in range(k + 2):
            if i == j: continue
            val = val * (mod + xs[i] - xs[j]) % mod
        
        saved_vals[i] = ys[i] * pow(val, -1, mod) % mod
    
    for i in range(k + 2):
        val = 0
        
        for j in range(k + 2):
            if i == j: continue
            val += saved_vals[j] * (mod + xs[i] - xs[j]) % mod
            val %= mod
        
        if val == 0:
            ans = 0
            for j in range(k + 2):
                if j != i:
                    A = ys[j]
                    B = 1
                    for v in range(k + 2):
                        if v != i and v != j:
                            A = A * (mod - xs[v]) % mod
                            B = B * (mod + xs[j] - xs[v]) % mod
                    ans += A * pow(B, -1, mod) % mod
                    ans %= mod
            print(ans)
            break